[Faith Chua] iP

boredcoco opened the pull request:
# DukePro
> Your companion throughout your busy life

### Tired of trying to juggle work, studies and extra-curriculars? Fret not, DukePro has you covered.

DukePro is:
- text-based
- user friendly
- and takes up **little** space on your laptop

To use DukePro, all you gotta do is:
1. download it from the repo [here](https://github.com/boredcoco/ip)
2. double click it
3. add your tasks
4. you're good to go!

### Best of all, DukePro is free for all!

What would you use DukePro for:
- [ ] managing deadlines
- [ ] managing appointments and events
- [ ] managing todos

### You can use DukePro to practice Java too!

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```
# Credits
I would like to give credit to [this SE Engineering textbook](https://se-education.org/se-book/ ) for teaching me to use GUI and JavaFX. A lot of my code was reused from there and adapted for this project. 

### Thank you for using DukePro, here is a cute hedgehog!

![This is an image of a hedgehog](https://i.pinimg.com/originals/98/ba/48/98ba48c230f378e064a02ec15c3b7227.jpg)

Commit:
Add Gradle support

boredcoco committed:
All levels

This repo was made after finishing ip in a separate repo, I noticed that I had cloned the repository from the previous year instead of this year's repository prior so I transferred and copied all the code I had written here. Thus there is only one commit

boredcoco committed:
Level-7

add data/tasklist.txt

boredcoco committed:
Level-7

add Decoder

boredcoco committed:
Level-8

boredcoco committed:
Merge pull request #1 from boredcoco/branch-Level-7

Branch level 7

boredcoco committed:
Merge pull request #2 from boredcoco/branch-Level-8

Level-8

boredcoco committed:
A-More-OOP

add Storage

boredcoco committed:
A-Packages

add Duke
add DukeExceptions
add DukeTasks

boredcoco committed:
Gradle

add gradle

boredcoco committed:
Merge pull request #3 from boredcoco/add-gradle-support

Gradle

boredcoco committed:
Revert "Merge pull request #3 from boredcoco/add-gradle-support"

This reverts commit 78fc8e557f970dacb8b468d144176e10fefcf4bc, reversing
changes made to a60844e070844080e12ac2e97a0769d41660c3d3.

boredcoco committed:
Merge pull request #4 from nus-cs2103-AY2223S1/add-gradle-support

Add Gradle support

boredcoco committed:
A-JUnit

add InteractTest.java

boredcoco committed:
A-JavaDoc

boredcoco committed:
A-CodingStandard

Update Decoder
Update Deadline
Update TasksManager

boredcoco committed:
Level-9

Update Decoder
Update Task
Update Interact
Update TasksManager

boredcoco committed:
Merge pull request #5 from boredcoco/branch-A-JavaDoc

A-JavaDoc

boredcoco committed:
Merge branch 'master' into branch-A-CodingStandard

boredcoco committed:
Merge pull request #6 from boredcoco/branch-A-CodingStandard

A-CodingStandard

boredcoco committed:
Merge branch 'master' into branch-Level-9

boredcoco committed:
Merge pull request #7 from boredcoco/branch-Level-9

Level-9

boredcoco committed:
Update Decoder.java

boredcoco committed:
Update all files

All files updated to reflect coding standards

boredcoco committed:
add DialogBox add DukePro add Launcher add DaDuke add DaUser

boredcoco committed:
Add DukeHandlers

Tasksmanagers and Interact only print to the console.

Due to GUI, this needs to change.

Thus Strings are returned instead of just printing to the console.

boredcoco committed:
add Javadocs

Javadocs was not updated prior.

Rewrote Javadocs in accordance with current updated functions and classes.

boredcoco committed:
Merge pull request #8 from boredcoco/branch-Level-10

Branch level 10

quannam0124 commented on src/main/java/Duke/DukePro.java:
```diff
@@ -0,0 +1,163 @@
+package Duke;
```
Overall your code does not violate any major coding standards so far! Great job!

quannam0124 commented on src/main/java/Duke/DukeTasks/Task.java:
```diff
@@ -0,0 +1,64 @@
+
+public class Task {
+    private String name;
+    private boolean finished = false;
```
Maybe you could change this to the name of this boolean
```suggestion
    private boolean isFinished = false;
```

quannam0124 commented on src/main/java/Duke/DukeHandlers/Storage.java:
```diff
@@ -0,0 +1,109 @@
+
+import Duke.DukeTasks.Task;
+
+import java.io.*;
```
Should this import be a wildcard import?

quannam0124 reviewed (commented):
Overall your code is quite clean and easy to read, except for a few minor coding standard violations. Good job!

boredcoco committed:
Add checkstyle.xml Add suppressions.xml

boredcoco committed:
Update all files in dukepro package

There were many checkstyle errors at first.

All dheckstyle errors have been resolved except for "unused javadoc tag" because every method should have a javadoc tag thus I did not delete any so called "unused java doc" tags

Ramanathan0908 commented on src/main/java/dukepro/DukePro.java:
```diff
@@ -0,0 +1,165 @@
+package dukepro;
+
+import dukepro.exceptions.DukeException;
+import dukepro.handlers.Interact;
+import javafx.application.Application;
+import javafx.scene.Scene;
+import javafx.scene.control.Button;
+import javafx.scene.control.Label;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.Region;
+import javafx.scene.layout.VBox;
+import javafx.stage.Stage;
```
A very minor nitpick but I would love to see some separation between imports from different packages for readability

Ramanathan0908 commented on src/main/java/dukepro/handlers/TasksManager.java:
```diff
@@ -0,0 +1,152 @@
+ * Class for TasksManager.
+ */
+public class TasksManager {
+    private static ArrayList<Task> tasks = new ArrayList<>();
```
Interesting to initialise the arraylist here, perhaps you could explain why you choose to initialise the arraylist here instead of the constructor?

Ramanathan0908 commented on src/main/java/dukepro/handlers/TasksManager.java:
```diff
@@ -0,0 +1,152 @@
+        //add to the tasks
+        tasks.add(task);
+        //write to file
+        storage.addTask(task);
```
Also curious to why you would implement the storage of tasks like this and not save the entire tasklist after each user command in the main program?

Ramanathan0908 commented on src/main/java/dukepro/DukePro.java:
```diff
@@ -0,0 +1,165 @@
+     * @return void.
+     */
+    @Override
+    public void start(Stage stage) {
```
I love how you explained different steps in your start function

Ramanathan0908 commented on src/test/java/InteractTest.java:
```diff
@@ -0,0 +1,54 @@
+import dukepro.exceptions.DukeException;
+import dukepro.handlers.Decoder;
+import dukepro.handlers.TasksManager;
+import dukepro.tasks.*;
```
There shouldn't be any wildcard imports

Ramanathan0908 reviewed (commented):
Great job overall!!! Just some minor nits to fix!

fongyj commented on src/main/java/dukepro/exceptions/BadFormatException.java:
```diff
@@ -0,0 +1,61 @@
+            return ret;
+        } else {
+            String ret = line + "\n Oh no, the format of a " + this.taskType + " is: " + this.format + "\n" + line;
+            return ret;
```
I think theres abit too much of repetition here inside the if else blocks. The logic implemented inside them seem to be rather similar. I think this portion can be refactored to take the common parts out to reduce repetition.

Also I feel that the variable "ret" can be named better as it might not be intuitive what ret is referring to.

fongyj commented on src/main/java/dukepro/exceptions/BadFormatException.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * Creates a BadFormatException.
+     *
+     * @param errorMessage the errormMessage.
```
I think "errormMessage" might be a typo. 

Also, while errorMessage is quite self explanatory, 
"@param errorMessage the errorMessage" seems rather redundant 
maybe "@param errorMessage the error message" would be nicer

fongyj commented on src/main/java/dukepro/exceptions/BadTaskOperationException.java:
```diff
@@ -0,0 +1,32 @@
+ * Class for BadTaskOperationException.
+ */
+public class BadTaskOperationException extends DukeException {
+    private static String line = "_______________________________________";
```
I believe the "line" variable might not be needed since this class extends DukeException which already have this variable

fongyj commented on src/main/java/dukepro/exceptions/EmptyDescException.java:
```diff
@@ -0,0 +1,30 @@
+ * Class for EmptyDescException.
+ */
+public class EmptyDescException extends DukeException {
+    private static String line = "_______________________________________";
```
I believe the "line" variable might not be needed since this class extends DukeException which already have this variable

fongyj commented on src/main/java/dukepro/handlers/Decoder.java:
```diff
@@ -0,0 +1,220 @@
+import dukepro.exceptions.BadFormatException;
+import dukepro.exceptions.BadTaskOperationException;
+import dukepro.exceptions.DukeException;
+import dukepro.exceptions.EmptyDescException;
```
Maybe a line break here would help make the import statements look more organised

fongyj commented on src/main/java/dukepro/handlers/Decoder.java:
```diff
@@ -0,0 +1,220 @@
+     * @throws DukeException if input is not proper.
+     */
+    public static Task handleTasks(String word) throws DukeException {
+        String[] splitted = word.split(" ", 2);
```
I think a better name for this variable would be "split" or "wordSplit". "splitted" feels more like a descriptive word rather than a name

fongyj commented on src/main/java/dukepro/handlers/Decoder.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * Returns is number is valid.
```
just happen to notice a typo here i think it should be "if"

fongyj commented on src/main/java/dukepro/handlers/Decoder.java:
```diff
@@ -0,0 +1,220 @@
+
+    /**
+     * Returns a string representing user's
+     * search input.
```
i think theres no need to line break here since the first line isnt that long. line break here might cause some confusion in reading the javadoc

fongyj commented on src/main/java/dukepro/handlers/Decoder.java:
```diff
@@ -0,0 +1,220 @@
+     * Returns is number is valid.
+     *
+     * @param num user input from command line.
+     * @return A boolean.
```
I think "@return A boolean." feels rather redundant since the method signature would already tell me the method returns a boolean. Maybe adding a short description like "@return true if number is valid" would fit better here even though it might feel abit repetitive since it can be inferred from your main description.

fongyj commented on src/main/java/dukepro/handlers/Interact.java:
```diff
@@ -0,0 +1,69 @@
+
+    /**
+     * Calls the task manager to do certain tasks
+     * depending on the command-line argument.
```
similar to what i mentioned above, i observe that you choose to linebreak in your javadoc after a small number of characters quite frequently. personally i feel this makes it harder to read as it breaks up a single sentence into many pieces. furthermore breaking the sentence up this early makes it seem like you apply a different line length limit for your javadoc and for your code below which can make it feel abit inconsistent and confusing.

i am not sure if theres a need to line break this early since it seems like there isnt that many characters in your first line. let me know if i missed any style requirement for this 👍

fongyj commented on src/main/java/dukepro/handlers/Storage.java:
```diff
@@ -0,0 +1,116 @@
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.ArrayList;
```
i think a line break here to organise the import statements would be neat

fongyj commented on src/main/java/dukepro/handlers/Storage.java:
```diff
@@ -0,0 +1,116 @@
+ * Class for Storage.
+ */
+public class Storage {
+    private String line = "_______________________________________";
```
i notice that this line variable seem to exist in many of your classes for the purpose of printing and it feels rather repetitive. i havent read your full implementation of dukebot in detail but maybe its possible to reduce this repetition by storing this line variable in a user interface class and getting the line variable from that class or calling a method from that class to handle printing for you.

furthermore it seems like in the Storage class the "line" variable is not used at all, i think this line can be removed too 👍

fongyj commented on src/main/java/dukepro/handlers/Storage.java:
```diff
@@ -0,0 +1,116 @@
+     *
+     * @return A Storage.
+     */
+    public Storage() { // creates new files if needed
```
i think this inline comment should be on the next line or nearer to your implementation of creating new files

fongyj commented on src/main/java/dukepro/handlers/Storage.java:
```diff
@@ -0,0 +1,116 @@
+     * the taskmanager.
+     *
+     * @param tasksManager The Tasksmanager to read data into.
+     * @return A boolean on whether reading was successful.
```
i like the description of the return value, it describes the meaning of the return boolean well.

fongyj commented on src/main/java/dukepro/tasks/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+     * @return A String in specific format.
+     */
+    @Override
+    public String fileForm() {
```
i think the method naming should be a verb so maybe something like "toFileForm" or "getFileForm"

fongyj reviewed (commented):
Other than some parts of the code that I feel can be improved on to match the code quality and standard set by the course, overall I think the code is well written 👍

boredcoco committed:
Update build.gradle

Change gradle build to reflect latest changes

After using JavaFX the main package is now renamed to "dukepro" rather than "duke".

Updating these changes in build.gradle tells shadowjar how to build the new .jar file

In order to do this, te archiveBaseName was changed from "duke" to "dukepro" to reflect the latest package and mainClassName is now "dukepro.Launcher"

boredcoco committed:
Add assert

Assert is not added in this project

Assert should be added so that the programmer can spot his or her errors more easily

Added some asserts in the dukepro.handlers.Decoder and dukepro.handlers.Interact.

The aforementioned classes are responsible for handling user input, and throwing errors when users provide bad input. The programmer may handle errors and parse inputs wrongly so these classes need assert to be added

boredcoco committed:
Update all classes to reflect java coding standard

boredcoco committed:
Merge pull request #10 from boredcoco/branch-A-Assertions

Add assert

boredcoco committed:
Merge branch 'master' into branch-A-CodeQuality

boredcoco committed:
Merge pull request #9 from boredcoco/branch-A-CodeQuality

Update all classes to reflect java coding standard

boredcoco committed:
Update displayed images

boredcoco committed:
Add Expense Add ExpenseManager Update Decoder Update Interact

boredcoco committed:
Update ExpenseManager Update Decoder Update Interact Update TasksManager

boredcoco committed:
Add StorableObjects Update Expense Update Task

A generic Storage class is required for  the expenses manager, so it is as extensible as possible.

Create a StorableaObjects class to abstract away storing details.

Internal implementation of Storage class needs to be changed.

boredcoco committed:
Add GeneralStorage Add Manager

Generic Managers and Storage classes are required to make the code more extensible

Replace TaskManager with a generic Manager and the Storage in TaskManager with a GeneralStorage.

Need to replace ExpenseManager with new generic Manager class.

boredcoco committed:
Update Interact

ExpenseManager field of Interact needed to be replaced with a generic Manager class.

Replace ExpenseManager in Interact with a generic Manager. Delete irrelevant classes like Storage, TaskManager, ExpenseManager

Need to implement a class or method that tabulates all expenses.

boredcoco committed:
Add ExpenseCalculator

Currently unable to tabulate total costs for all expenses

Add ExpenseCalculator class to support this functionality.

More specific functions can be a part of ExpenseCalculator in future, such as date specific functions etc.

boredcoco committed:
Update ExpenseCalculator

Need to update ExpenseCalculator so support action where user wants to know how much was spent in a day.

Add spentDay method to ExpenseCalculator so that user can find how much he / she has spent on a particular day.

User can now find out how much has been spent on a particular day by writing "spentDate" followed by the Date in YYYY-MM-DD format into the GUI.

boredcoco committed:
Update Javadocs

Javadocs was not updated to reflect the latest changes

Update Javadocs to reflect new functions and classes.

The code passes the checkstyle tests.

boredcoco committed:
Add TaskFunction

Not all StorableObjects require a "done" boolean field. Expenses are not "done".

Rather than having a "markAsDone" method in Manager, use the operationOnList method of the Manager class to handle the operation

May need to revamp updated code to be more OOP

boredcoco committed:
Update README.md

boredcoco committed:
Update JUnit tests

boredcoco committed:
Merge branch 'master' of https://github.com/boredcoco/ip

boredcoco committed:
Update DukePro Update Interact

Some methods and classes were not following the Java Coding Standard

Create new methods to apply SLAP and create better Header comments

boredcoco committed:
Add UI.png

boredcoco committed:
Add User Guide

boredcoco committed:
Update Expense Update Manager

The dates displayed was in YYYY-MM-DD format which was not that readable

Reformat existing local dates to show the Month (eg. Jan) followed by the year (eg. 2022) then the day (eg. 13). An example: Jan 2022 13

boredcoco committed:
Update README.md

boredcoco committed:
Update DukePro Update Interact

boredcoco committed:
Update Readme.md Update DukePro.java

boredcoco committed:
Update Expense

Expense previously did not accept an amount that was a float or double or had a decimal.

Change the amount field of an Expense to be a double instead of an expense.

boredcoco committed:
Update BadFormatException Update  BadTaskOperationException Update DukeException Update Decoder

boredcoco committed:
Update README.md Update Decoder Update Interact Update TaskFunction Update Task Update InteractTest

Previously this app did not have any way to mark a task as undone

Change the command of marking from "done" to "mark" and include an "unmark" functionality

boredcoco committed:
Update Decoder

If the Task to be deleted was 0 or any negative number or if the expense to be deleted was 0 or any negative number, no command could be sent and an error would occur.

Amend Decoder class to account for such cases

boredcoco committed:
Update README.md

boredcoco committed:
Update README.md
